File: landrop-main/sender/desktop/src/protocol/secure-channel.test.ts
import { describe, expect, it } from "vitest";

import {
  SECURE_CHANNEL_ESTABLISHED_EVENT,
  CONNECTION_LOST_EVENT,
  SECURE_CHANNEL_OUTCOMES,
  SecureChannelAttempt,
  secureChannelFailure,
} from "./secure-channel";

const fixtures: Record<string, unknown> = require(
  "../../../../protocol/fixtures/model-cases.json",
);

const secureChannel = fixtures.secure_channel as {
  outcomes: string[];
  failure_mapping: { outcome: string; transport_kind: string; message: string }[];
  event_mapping: { outcome: string; event: string }[];
};

describe("phase 05 — secure channel lifecycle foundation", () => {
  it("matches the fixture outcome vocabulary", () => {
    expect(SECURE_CHANNEL_OUTCOMES).toEqual(
      ["established", "handshake_failure", "validation_failure", "timeout", "closed"],
    );
    expect(secureChannel.outcomes).toEqual(SECURE_CHANNEL_OUTCOMES);
  });

  it("handshake phase emits no delivery event (machine stays CONNECTING)", () => {
    const channel = new SecureChannelAttempt();
    expect(channel.begin()).toBeNull();
    expect(channel.currentPhase).toBe("tls");
  });

  it("establishes the secure channel and emits secure_channel_established", () => {
    const channel = new SecureChannelAttempt();
    channel.begin();
    expect(channel.complete("established")).toBe("secure_channel_established");
    expect(channel.currentPhase).toBe("established");
    expect(channel.events()).toEqual([SECURE_CHANNEL_ESTABLISHED_EVENT]);
  });

  it("is finalized after establishment: repeated completion is a no-op", () => {
    const channel = new SecureChannelAttempt();
    channel.begin();
    channel.complete("established");
    expect(channel.complete("established")).toBeNull();
    expect(channel.complete("closed")).toBeNull();
  });

  it("maps every TLS failure to a transport failure per the fixture", () => {
    for (const row of secureChannel.failure_mapping) {
      const outcome = row.outcome;
      const failure = secureChannelFailure(
        outcome as "handshake_failure" | "validation_failure" | "timeout" | "closed",
      );
      expect(failure, row.outcome).toEqual({ kind: row.transport_kind, message: row.message });
    }
  });

  it("failing the handshake emits connection_lost once and keeps the failure", () => {
    for (const outcome of ["handshake_failure", "validation_failure", "timeout", "closed"] as const) {
      const channel = new SecureChannelAttempt();
      channel.begin();
      expect(channel.complete(outcome)).toBe("connection_lost");
      expect(channel.currentPhase).toBe("failed");
      expect(channel.lastFailure).toEqual(secureChannelFailure(outcome));
      // Deterministic failure propagation/finalization.
      expect(channel.complete(outcome)).toBeNull();
      expect(channel.events()).toEqual([CONNECTION_LOST_EVENT]);
    }
  });

  it("aborting the handshake emits connection_lost once and is idempotent", () => {
    const channel = new SecureChannelAttempt();
    channel.begin();
    expect(channel.close()).toBe("connection_lost");
    expect(channel.currentPhase).toBe("closed");
    expect(channel.close()).toBeNull();
    expect(channel.events()).toEqual([CONNECTION_LOST_EVENT]);
  });

  it("reproduces the fixture outcome-to-event mapping deterministically", () => {
    for (const row of secureChannel.event_mapping) {
      const channel = new SecureChannelAttempt();
      channel.begin();
      const event = channel.complete(row.outcome as never);
      expect(event, row.outcome).toBe(row.event);
      expect(channel.events(), row.outcome).toEqual([row.event]);
    }
  });
});

File: landrop-main/receiver/android/app/src/test/java/com/landrop/receiver/protocol/SecureChannelTest.kt
package com.landrop.receiver.protocol

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class SecureChannelTest {

    @Test
    fun matchesFixtureOutcomeVocabulary() {
        assertEquals(
            listOf(
                SecureChannelOutcome.ESTABLISHED,
                SecureChannelOutcome.HANDSHAKE_FAILURE,
                SecureChannelOutcome.VALIDATION_FAILURE,
                SecureChannelOutcome.TIMEOUT,
                SecureChannelOutcome.CLOSED,
            ),
            SECURE_CHANNEL_OUTCOMES,
        )
    }

    @Test
    fun handshakePhaseEmitsNoDeliveryEvent() {
        val channel = SecureChannelAttempt()
        channel.begin()
        assertEquals(SecureChannelPhase.TLS, channel.phase)
    }

    @Test
    fun establishesSecureChannelAndEmitsEvent() {
        val channel = SecureChannelAttempt()
        channel.begin()
        assertEquals(DeliveryEvent.SECURE_CHANNEL_ESTABLISHED, channel.complete(SecureChannelOutcome.ESTABLISHED))
        assertEquals(SecureChannelPhase.ESTABLISHED, channel.phase)
        assertEquals(listOf(DeliveryEvent.SECURE_CHANNEL_ESTABLISHED), channel.events())
    }

    @Test
    fun isFinalizedAfterEstablishment() {
        val channel = SecureChannelAttempt()
        channel.begin()
        channel.complete(SecureChannelOutcome.ESTABLISHED)
        assertNull(channel.complete(SecureChannelOutcome.ESTABLISHED))
        assertNull(channel.complete(SecureChannelOutcome.CLOSED))
    }

    @Test
    fun mapsEveryTlsFailurePerFixture() {
        val cases = listOf(
            SecureChannelOutcome.HANDSHAKE_FAILURE to TransportFailure(TransportErrorKind.IO_ERROR, "TLS handshake failed"),
            SecureChannelOutcome.VALIDATION_FAILURE to TransportFailure(
                TransportErrorKind.IO_ERROR,
                "certificate validation failed (trust model undefined)",
            ),
            SecureChannelOutcome.TIMEOUT to TransportFailure(TransportErrorKind.TIMEOUT, "TLS handshake timed out"),
            SecureChannelOutcome.CLOSED to TransportFailure(
                TransportErrorKind.CONNECTION_LOST,
                "connection closed during TLS handshake",
            ),
        )
        for ((outcome, expected) in cases) {
            assertEquals(expected, secureChannelFailure(outcome))
        }
    }

    @Test
    fun failingHandshakeEmitsConnectionLostOnceAndKeepsFailure() {
        for (outcome in listOf(
            SecureChannelOutcome.HANDSHAKE_FAILURE,
            SecureChannelOutcome.VALIDATION_FAILURE,
            SecureChannelOutcome.TIMEOUT,
            SecureChannelOutcome.CLOSED,
        )) {
            val channel = SecureChannelAttempt()
            channel.begin()
            assertEquals(DeliveryEvent.CONNECTION_LOST, channel.complete(outcome))
            assertEquals(SecureChannelPhase.FAILED, channel.phase)
            assertEquals(secureChannelFailure(outcome), channel.lastFailure)
            assertNull(channel.complete(outcome))
            assertEquals(listOf(DeliveryEvent.CONNECTION_LOST), channel.events())
        }
    }

    @Test
    fun abortingHandshakeEmitsConnectionLostOnceAndIsIdempotent() {
        val channel = SecureChannelAttempt()
        channel.begin()
        assertEquals(DeliveryEvent.CONNECTION_LOST, channel.close())
        assertEquals(SecureChannelPhase.CLOSED, channel.phase)
        assertNull(channel.close())
        assertEquals(listOf(DeliveryEvent.CONNECTION_LOST), channel.events())
    }

    @Test
    fun reproducesFixtureOutcomeToEventMapping() {
        val cases = listOf(
            SecureChannelOutcome.ESTABLISHED to DeliveryEvent.SECURE_CHANNEL_ESTABLISHED,
            SecureChannelOutcome.HANDSHAKE_FAILURE to DeliveryEvent.CONNECTION_LOST,
            SecureChannelOutcome.VALIDATION_FAILURE to DeliveryEvent.CONNECTION_LOST,
            SecureChannelOutcome.TIMEOUT to DeliveryEvent.CONNECTION_LOST,
            SecureChannelOutcome.CLOSED to DeliveryEvent.CONNECTION_LOST,
        )
        for ((outcome, event) in cases) {
            val channel = SecureChannelAttempt()
            channel.begin()
            assertEquals(event, channel.complete(outcome))
            assertEquals(listOf(event), channel.events())
        }
    }
}